`OutOfMemoryException` on 2.0.0

Looks like we have an `OutOfMemoryException` followed by an `OutOfDirectMemoryError`.
I then find that it's definitely leaked memory because queries that weren't running out of memory before now start to.

Just capturing the stack trace for now:

<details><summary>Stack trace</summary>

```
19:11:13 | ERROR xtdb.pgwire | Uncaught exception processing message
xtdb.error.Fault: Failure allocating buffer.
	at xtdb.error$fault.invokeStatic(error.clj:90)
	at xtdb.error$fault.invoke(error.clj:82)
	at xtdb.error$eval752$fn__753.invoke(error.clj:204)
	at xtdb.error$eval671$fn__672$G__662__679.invoke(error.clj:14)
	at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1395)
	at xtdb.pgwire$handle_msg.invoke(pgwire.clj:1391)
	at xtdb.pgwire$conn_loop.invokeStatic(pgwire.clj:1444)
	at xtdb.pgwire$conn_loop.invoke(pgwire.clj:1415)
	at xtdb.pgwire$connect.invokeStatic(pgwire.clj:1494)
	at xtdb.pgwire$connect.invoke(pgwire.clj:1454)
	at xtdb.pgwire$accept_loop$fn__27775$fn__27776$fn__27777.invoke(pgwire.clj:1539)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.apache.arrow.memory.OutOfMemoryException: Failure allocating buffer.
	at io.netty.buffer.PooledByteBufAllocatorL.allocate(PooledByteBufAllocatorL.java:62)
	at org.apache.arrow.memory.netty.NettyAllocationManager.<init>(NettyAllocationManager.java:74)
	at org.apache.arrow.memory.netty.NettyAllocationManager.<init>(NettyAllocationManager.java:81)
	at org.apache.arrow.memory.netty.NettyAllocationManager$1.create(NettyAllocationManager.java:38)
	at org.apache.arrow.memory.BaseAllocator.newAllocationManager(BaseAllocator.java:376)
	at org.apache.arrow.memory.BaseAllocator.newAllocationManager(BaseAllocator.java:372)
	at org.apache.arrow.memory.BaseAllocator.bufferWithoutReservation(BaseAllocator.java:357)
	at org.apache.arrow.memory.BaseAllocator.buffer(BaseAllocator.java:336)
	at org.apache.arrow.memory.BaseAllocator.buffer(BaseAllocator.java:298)
	at org.apache.arrow.vector.BaseVariableWidthVector.reallocDataBuffer(BaseVariableWidthVector.java:579)
	at org.apache.arrow.vector.BaseVariableWidthVector.handleSafe(BaseVariableWidthVector.java:1350)
	at org.apache.arrow.vector.BaseVariableWidthVector.copyFromSafe(BaseVariableWidthVector.java:1449)
	at xtdb.vector.ScalarVectorWriter.rowCopier$lambda$1(FieldVectorWriters.kt:86)
	at xtdb.vector.IndirectVectorReader.lambda$rowCopier$2(IndirectVectorReader.java:184)
	at xtdb.vector.IndirectVectorReader.copyTo(IndirectVectorReader.java:173)
	at xtdb.vector.IVectorReader.copy(IVectorReader.kt:32)
	at xtdb.vector.RenamedVectorReader.copy(RenamedVectorReader.java:160)
	at xtdb.vector.RelationReader.copy$lambda$4(RelationReader.kt:35)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1939)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)
	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622)
	at java.base/java.util.stream.ReferencePipeline.toList(ReferencePipeline.java:627)
	at xtdb.vector.RelationReader.from(RelationReader.kt:29)
	at xtdb.vector.RelationReader.copy(RelationReader.kt:35)
	at xtdb.operator.join.JoinCursor$fn$reify__23918.accept(join.clj:350)
	at xtdb.operator.ProjectCursor.tryAdvance$lambda$3(ProjectCursor.kt:27)
	at xtdb.operator.ProjectCursor.tryAdvance$lambda$4(ProjectCursor.kt:17)
	at xtdb.operator.ProjectCursor.tryAdvance$lambda$3(ProjectCursor.kt:27)
	at xtdb.operator.ProjectCursor.tryAdvance$lambda$4(ProjectCursor.kt:17)
	at xtdb.operator.join.JoinCursor$fn$reify__23918.accept(join.clj:353)
	at xtdb.operator.ProjectCursor.tryAdvance$lambda$3(ProjectCursor.kt:27)
	at xtdb.operator.ProjectCursor.tryAdvance$lambda$4(ProjectCursor.kt:17)
	at xtdb.operator.rename.RenameCursor$fn__24625.invoke(rename.clj:39)
	at clojure.lang.FnInvokers.invokeOO(FnInvokers.java:247)
	at xtdb.operator.ProjectCursor.tryAdvance$lambda$3(ProjectCursor.kt:27)
	at xtdb.operator.ProjectCursor.tryAdvance$lambda$4(ProjectCursor.kt:17)
	at xtdb.coalesce.CoalescingCursor$reify__24705.accept(coalesce.clj:34)
	at xtdb.operator.SelectCursor.tryAdvance$lambda$0(SelectCursor.kt:22)
	at xtdb.operator.SelectCursor.tryAdvance$lambda$1(SelectCursor.kt:19)
	at xtdb.operator.table.TableCursor$fn__25000.invoke(table.clj:37)
	at xtdb.operator.table.TableCursor.tryAdvance(table.clj:35)
	at xtdb.operator.SelectCursor.tryAdvance(SelectCursor.kt:19)
	at xtdb.coalesce.CoalescingCursor.tryAdvance(coalesce.clj:24)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.rename.RenameCursor.tryAdvance(rename.clj:31)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.join.JoinCursor$fn__23913.invoke(join.clj:345)
	at xtdb.operator.join.JoinCursor.tryAdvance(join.clj:338)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.join.JoinCursor$fn__23913.invoke(join.clj:345)
	at xtdb.operator.join.JoinCursor.tryAdvance(join.clj:338)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.join.JoinCursor$fn__23913.invoke(join.clj:345)
	at xtdb.operator.join.JoinCursor.tryAdvance(join.clj:338)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.join.JoinCursor$fn__23913.invoke(join.clj:345)
	at xtdb.operator.join.JoinCursor.tryAdvance(join.clj:338)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.join.JoinCursor$fn__23913.invoke(join.clj:345)
	at xtdb.operator.join.JoinCursor.tryAdvance(join.clj:338)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.join.JoinCursor$fn__23913.invoke(join.clj:345)
	at xtdb.operator.join.JoinCursor.tryAdvance(join.clj:338)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.join.JoinCursor$fn__23913.invoke(join.clj:345)
	at xtdb.operator.join.JoinCursor.tryAdvance(join.clj:338)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:332)
	at xtdb.operator.group_by.GroupByCursor$fn__15482.invoke(group_by.clj:594)
	at xtdb.operator.group_by.GroupByCursor.tryAdvance(group_by.clj:593)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.rename.RenameCursor.tryAdvance(rename.clj:31)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:332)
	at xtdb.operator.join$build_phase.invokeStatic(join.clj:165)
	at xtdb.operator.join$build_phase.invoke(join.clj:164)
	at xtdb.operator.join.JoinCursor.tryAdvance(join.clj:334)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:332)
	at xtdb.operator.group_by.GroupByCursor$fn__15482.invoke(group_by.clj:594)
	at xtdb.operator.group_by.GroupByCursor.tryAdvance(group_by.clj:593)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.operator.order_by.OrderByCursor.tryAdvance(order_by.clj:233)
	at xtdb.operator.ProjectCursor.tryAdvance(ProjectCursor.kt:17)
	at xtdb.query$wrap_cursor$reify__25546.tryAdvance(query.clj:70)
	at xtdb.pgwire$cmd_exec_query.invokeStatic(pgwire.clj:1166)
	at xtdb.pgwire$cmd_exec_query.invoke(pgwire.clj:1152)
	at xtdb.pgwire$execute_portal.invokeStatic(pgwire.clj:1237)
	at xtdb.pgwire$execute_portal.invoke(pgwire.clj:1202)
	at xtdb.pgwire$eval27639$fn__27642.invoke(pgwire.clj:1271)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1398)
	... 12 common frames omitted
Caused by: io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 67108864 byte(s) of direct memory (used: 6232250119, max: 6291456000)
	at io.netty.util.internal.PlatformDependent.incrementMemoryCounter(PlatformDependent.java:822)
	at io.netty.util.internal.PlatformDependent.allocateDirectNoCleaner(PlatformDependent.java:751)
	at io.netty.buffer.UnpooledUnsafeNoCleanerDirectByteBuf.allocateDirect(UnpooledUnsafeNoCleanerDirectByteBuf.java:30)
	at io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeNoCleanerDirectByteBuf.allocateDirect(UnpooledByteBufAllocator.java:186)
	at io.netty.buffer.UnpooledDirectByteBuf.<init>(UnpooledDirectByteBuf.java:64)
	at io.netty.buffer.UnpooledUnsafeDirectByteBuf.<init>(UnpooledUnsafeDirectByteBuf.java:41)
	at io.netty.buffer.UnpooledUnsafeNoCleanerDirectByteBuf.<init>(UnpooledUnsafeNoCleanerDirectByteBuf.java:25)
	at io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeNoCleanerDirectByteBuf.<init>(UnpooledByteBufAllocator.java:181)
	at io.netty.buffer.UnpooledByteBufAllocator.newDirectBuffer(UnpooledByteBufAllocator.java:91)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:188)
	at io.netty.buffer.PooledByteBufAllocatorL$InnerAllocator.newDirectBufferL(PooledByteBufAllocatorL.java:163)
	at io.netty.buffer.PooledByteBufAllocatorL$InnerAllocator.directBuffer(PooledByteBufAllocatorL.java:206)
	at io.netty.buffer.PooledByteBufAllocatorL.allocate(PooledByteBufAllocatorL.java:53)
	... 116 common frames omitted
19:11:13 | ERROR o.apache.arrow.memory.BaseAllocator | Memory was leaked by query. Memory leaked: (143130624)
Allocator(BoundQuery/openCursor) 0/143130624/433238544/9223372036854775807 (res/actual/peak/limit)
19:11:13 | ERROR xtdb.pgwire | Uncaught exception processing message
xtdb.error.Fault: Memory was leaked by query. Memory leaked: (143130624)
Allocator(BoundQuery/openCursor) 0/143130624/433238544/9223372036854775807 (res/actual/peak/limit)
	at xtdb.error$fault.invokeStatic(error.clj:90)
	at xtdb.error$fault.invoke(error.clj:82)
	at xtdb.error$eval752$fn__753.invoke(error.clj:204)
	at xtdb.error$eval671$fn__672$G__662__679.invoke(error.clj:14)
	at xtdb.pgwire$send_ex.invokeStatic(pgwire.clj:254)
	at xtdb.pgwire$send_ex.invoke(pgwire.clj:250)
	at xtdb.pgwire$eval27273$fn__27275$fn__27278.invoke(pgwire.clj:586)
	at xtdb.pgwire$eval27273$fn__27275.invoke(pgwire.clj:579)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1398)
	at xtdb.pgwire$handle_msg.invoke(pgwire.clj:1391)
	at xtdb.pgwire$conn_loop.invokeStatic(pgwire.clj:1444)
	at xtdb.pgwire$conn_loop.invoke(pgwire.clj:1415)
	at xtdb.pgwire$connect.invokeStatic(pgwire.clj:1494)
	at xtdb.pgwire$connect.invoke(pgwire.clj:1454)
	at xtdb.pgwire$accept_loop$fn__27775$fn__27776$fn__27777.invoke(pgwire.clj:1539)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalStateException: Memory was leaked by query. Memory leaked: (143130624)
Allocator(BoundQuery/openCursor) 0/143130624/433238544/9223372036854775807 (res/actual/peak/limit)
	at org.apache.arrow.memory.BaseAllocator.close(BaseAllocator.java:504)
	at xtdb.util$close.invokeStatic(util.clj:52)
	at xtdb.util$close.invoke(util.clj:35)
	at xtdb.query$wrap_cursor$reify__25546.close(query.clj:83)
	at xtdb.util$close.invokeStatic(util.clj:52)
	at xtdb.util$close.invoke(util.clj:35)
	at xtdb.pgwire$close_portal.invokeStatic(pgwire.clj:389)
	at xtdb.pgwire$close_portal.invoke(pgwire.clj:385)
	at xtdb.pgwire$close_all_portals.invokeStatic(pgwire.clj:395)
	at xtdb.pgwire$close_all_portals.invoke(pgwire.clj:393)
	at xtdb.pgwire$eval27273$fn__27275$fn__27278.invoke(pgwire.clj:584)
	... 15 common frames omitted
19:11:13 | ERROR xtdb.pgwire | Uncaught exception processing message
xtdb.error.Fault: RefCnt has gone negative
	at xtdb.error$fault.invokeStatic(error.clj:90)
	at xtdb.error$fault.invoke(error.clj:82)
	at xtdb.error$eval752$fn__753.invoke(error.clj:204)
	at xtdb.error$eval671$fn__672$G__662__679.invoke(error.clj:14)
	at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1395)
	at xtdb.pgwire$handle_msg.invoke(pgwire.clj:1391)
	at xtdb.pgwire$conn_loop.invokeStatic(pgwire.clj:1444)
	at xtdb.pgwire$conn_loop.invoke(pgwire.clj:1415)
	at xtdb.pgwire$connect.invokeStatic(pgwire.clj:1494)
	at xtdb.pgwire$connect.invoke(pgwire.clj:1454)
	at xtdb.pgwire$accept_loop$fn__27775$fn__27776$fn__27777.invoke(pgwire.clj:1539)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalStateException: RefCnt has gone negative
	at org.apache.arrow.util.Preconditions.checkState(Preconditions.java:419)
	at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:128)
	at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:104)
	at org.apache.arrow.memory.ArrowBuf.close(ArrowBuf.java:1032)
	at xtdb.arrow.ExtensibleBuffer.close(Buffers.kt:161)
	at xtdb.arrow.FixedWidthVector.close(FixedWidthVector.kt:154)
	at xtdb.util.IOUtilKt.closeAll(IOUtil.kt:27)
	at xtdb.arrow.RelationReader.close(RelationReader.kt:26)
	at xtdb.util$close.invokeStatic(util.clj:52)
	at xtdb.util$close.invoke(util.clj:35)
	at xtdb.query$wrap_cursor$reify__25546.close(query.clj:82)
	at xtdb.util$close.invokeStatic(util.clj:52)
	at xtdb.util$close.invoke(util.clj:35)
	at xtdb.pgwire$close_portal.invokeStatic(pgwire.clj:389)
	at xtdb.pgwire$close_portal.invoke(pgwire.clj:385)
	at xtdb.pgwire$eval27536$fn__27539.invoke(pgwire.clj:1045)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1398)
	... 12 common frames omitted
19:11:13 | ERROR xtdb.pgwire | Uncaught exception processing message
xtdb.error.Fault: RefCnt has gone negative
	at xtdb.error$fault.invokeStatic(error.clj:90)
	at xtdb.error$fault.invoke(error.clj:82)
	at xtdb.error$eval752$fn__753.invoke(error.clj:204)
	at xtdb.error$eval671$fn__672$G__662__679.invoke(error.clj:14)
	at xtdb.pgwire$send_ex.invokeStatic(pgwire.clj:254)
	at xtdb.pgwire$send_ex.invoke(pgwire.clj:250)
	at xtdb.pgwire$eval27273$fn__27275$fn__27278.invoke(pgwire.clj:586)
	at xtdb.pgwire$eval27273$fn__27275.invoke(pgwire.clj:579)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1398)
	at xtdb.pgwire$handle_msg.invoke(pgwire.clj:1391)
	at xtdb.pgwire$conn_loop.invokeStatic(pgwire.clj:1444)
	at xtdb.pgwire$conn_loop.invoke(pgwire.clj:1415)
	at xtdb.pgwire$connect.invokeStatic(pgwire.clj:1494)
	at xtdb.pgwire$connect.invoke(pgwire.clj:1454)
	at xtdb.pgwire$accept_loop$fn__27775$fn__27776$fn__27777.invoke(pgwire.clj:1539)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalStateException: RefCnt has gone negative
	at org.apache.arrow.util.Preconditions.checkState(Preconditions.java:419)
	at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:128)
	at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:104)
	at org.apache.arrow.memory.ArrowBuf.close(ArrowBuf.java:1032)
	at xtdb.arrow.ExtensibleBuffer.close(Buffers.kt:161)
	at xtdb.arrow.FixedWidthVector.close(FixedWidthVector.kt:154)
	at xtdb.util.IOUtilKt.closeAll(IOUtil.kt:27)
	at xtdb.arrow.RelationReader.close(RelationReader.kt:26)
	at xtdb.util$close.invokeStatic(util.clj:52)
	at xtdb.util$close.invoke(util.clj:35)
	at xtdb.query$wrap_cursor$reify__25546.close(query.clj:82)
	at xtdb.util$close.invokeStatic(util.clj:52)
	at xtdb.util$close.invoke(util.clj:35)
	at xtdb.pgwire$close_portal.invokeStatic(pgwire.clj:389)
	at xtdb.pgwire$close_portal.invoke(pgwire.clj:385)
	at xtdb.pgwire$close_all_portals.invokeStatic(pgwire.clj:395)
	at xtdb.pgwire$close_all_portals.invoke(pgwire.clj:393)
	at xtdb.pgwire$eval27273$fn__27275$fn__27278.invoke(pgwire.clj:584)
	... 15 common frames omitted
19:11:13 | ERROR xtdb.pgwire | Uncaught exception processing message
xtdb.error.Fault: RefCnt has gone negative
	at xtdb.error$fault.invokeStatic(error.clj:90)
	at xtdb.error$fault.invoke(error.clj:82)
	at xtdb.error$eval752$fn__753.invoke(error.clj:204)
	at xtdb.error$eval671$fn__672$G__662__679.invoke(error.clj:14)
	at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1395)
	at xtdb.pgwire$handle_msg.invoke(pgwire.clj:1391)
	at xtdb.pgwire$conn_loop.invokeStatic(pgwire.clj:1444)
	at xtdb.pgwire$conn_loop.invoke(pgwire.clj:1415)
	at xtdb.pgwire$connect.invokeStatic(pgwire.clj:1494)
	at xtdb.pgwire$connect.invoke(pgwire.clj:1454)
	at xtdb.pgwire$accept_loop$fn__27775$fn__27776$fn__27777.invoke(pgwire.clj:1539)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalStateException: RefCnt has gone negative
	at org.apache.arrow.util.Preconditions.checkState(Preconditions.java:419)
	at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:128)
	at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:104)
	at org.apache.arrow.memory.ArrowBuf.close(ArrowBuf.java:1032)
	at xtdb.arrow.ExtensibleBuffer.close(Buffers.kt:161)
	at xtdb.arrow.FixedWidthVector.close(FixedWidthVector.kt:154)
	at xtdb.util.IOUtilKt.closeAll(IOUtil.kt:27)
	at xtdb.arrow.RelationReader.close(RelationReader.kt:26)
	at xtdb.util$close.invokeStatic(util.clj:52)
	at xtdb.util$close.invoke(util.clj:35)
	at xtdb.query$wrap_cursor$reify__25546.close(query.clj:82)
	at xtdb.util$close.invokeStatic(util.clj:52)
	at xtdb.util$close.invoke(util.clj:35)
	at xtdb.pgwire$close_portal.invokeStatic(pgwire.clj:389)
	at xtdb.pgwire$close_portal.invoke(pgwire.clj:385)
	at xtdb.pgwire$eval27536$fn__27539.invoke(pgwire.clj:1045)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1398)
	... 12 common frames omitted
19:11:13 | ERROR xtdb.pgwire | Uncaught exception processing message
xtdb.error.Fault: RefCnt has gone negative
	at xtdb.error$fault.invokeStatic(error.clj:90)
	at xtdb.error$fault.invoke(error.clj:82)
	at xtdb.error$eval752$fn__753.invoke(error.clj:204)
	at xtdb.error$eval671$fn__672$G__662__679.invoke(error.clj:14)
	at xtdb.pgwire$send_ex.invokeStatic(pgwire.clj:254)
	at xtdb.pgwire$send_ex.invoke(pgwire.clj:250)
	at xtdb.pgwire$eval27273$fn__27275$fn__27278.invoke(pgwire.clj:586)
	at xtdb.pgwire$eval27273$fn__27275.invoke(pgwire.clj:579)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1398)
	at xtdb.pgwire$handle_msg.invoke(pgwire.clj:1391)
	at xtdb.pgwire$conn_loop.invokeStatic(pgwire.clj:1444)
	at xtdb.pgwire$conn_loop.invoke(pgwire.clj:1415)
	at xtdb.pgwire$connect.invokeStatic(pgwire.clj:1494)
	at xtdb.pgwire$connect.invoke(pgwire.clj:1454)
	at xtdb.pgwire$accept_loop$fn__27775$fn__27776$fn__27777.invoke(pgwire.clj:1539)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalStateException: RefCnt has gone negative
	at org.apache.arrow.util.Preconditions.checkState(Preconditions.java:419)
	at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:128)
	at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:104)
	at org.apache.arrow.memory.ArrowBuf.close(ArrowBuf.java:1032)
	at xtdb.arrow.ExtensibleBuffer.close(Buffers.kt:161)
	at xtdb.arrow.FixedWidthVector.close(FixedWidthVector.kt:154)
	at xtdb.util.IOUtilKt.closeAll(IOUtil.kt:27)
	at xtdb.arrow.RelationReader.close(RelationReader.kt:26)
	at xtdb.util$close.invokeStatic(util.clj:52)
	at xtdb.util$close.invoke(util.clj:35)
	at xtdb.query$wrap_cursor$reify__25546.close(query.clj:82)
	at xtdb.util$close.invokeStatic(util.clj:52)
	at xtdb.util$close.invoke(util.clj:35)
	at xtdb.pgwire$close_portal.invokeStatic(pgwire.clj:389)
	at xtdb.pgwire$close_portal.invoke(pgwire.clj:385)
	at xtdb.pgwire$close_all_portals.invokeStatic(pgwire.clj:395)
	at xtdb.pgwire$close_all_portals.invoke(pgwire.clj:393)
	at xtdb.pgwire$eval27273$fn__27275$fn__27278.invoke(pgwire.clj:584)
	... 15 common frames omitted
19:11:13 | ERROR xtdb.pgwire | Uncaught exception processing message
xtdb.error.Fault: RefCnt has gone negative
	at xtdb.error$fault.invokeStatic(error.clj:90)
	at xtdb.error$fault.invoke(error.clj:82)
	at xtdb.error$eval752$fn__753.invoke(error.clj:204)
	at xtdb.error$eval671$fn__672$G__662__679.invoke(error.clj:14)
	at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1395)
	at xtdb.pgwire$handle_msg.invoke(pgwire.clj:1391)
	at xtdb.pgwire$conn_loop.invokeStatic(pgwire.clj:1444)
	at xtdb.pgwire$conn_loop.invoke(pgwire.clj:1415)
	at xtdb.pgwire$connect.invokeStatic(pgwire.clj:1494)
	at xtdb.pgwire$connect.invoke(pgwire.clj:1454)
	at xtdb.pgwire$accept_loop$fn__27775$fn__27776$fn__27777.invoke(pgwire.clj:1539)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.IllegalStateException: RefCnt has gone negative
	at org.apache.arrow.util.Preconditions.checkState(Preconditions.java:419)
	at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:128)
	at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:104)
	at org.apache.arrow.memory.ArrowBuf.close(ArrowBuf.java:1032)
	at xtdb.arrow.ExtensibleBuffer.close(Buffers.kt:161)
	at xtdb.arrow.FixedWidthVector.close(FixedWidthVector.kt:154)
	at xtdb.util.IOUtilKt.closeAll(IOUtil.kt:27)
	at xtdb.arrow.RelationReader.close(RelationReader.kt:26)
	at xtdb.util$close.invokeStatic(util.clj:52)
	at xtdb.util$close.invoke(util.clj:35)
	at xtdb.query$wrap_cursor$reify__25546.close(query.clj:82)
	at xtdb.util$close.invokeStatic(util.clj:52)
	at xtdb.util$close.invoke(util.clj:35)
	at xtdb.pgwire$close_portal.invokeStatic(pgwire.clj:389)
	at xtdb.pgwire$close_portal.invoke(pgwire.clj:385)
	at xtdb.pgwire$close_all_portals.invokeStatic(pgwire.clj:395)
	at xtdb.pgwire$close_all_portals.invoke(pgwire.clj:393)
	at xtdb.pgwire$eval27168$fn__27170.invoke(pgwire.clj:418)
	at clojure.lang.MultiFn.invoke(MultiFn.java:234)
	at xtdb.pgwire$handle_msg.invokeStatic(pgwire.clj:1398)
	... 12 common frames omitted
```

</details>

--- a/core/src/main/java/xtdb/vector/IndirectVectorReader.java
+++ b/core/src/main/java/xtdb/vector/IndirectVectorReader.java
@@ -1,5 +1,6 @@
 package xtdb.vector;
 
+import org.apache.arrow.memory.BufferAllocator;
 import org.apache.arrow.memory.util.ArrowBufPointer;
 import org.apache.arrow.vector.ValueVector;
 import org.apache.arrow.vector.types.pojo.Field;
@@ -193,6 +194,11 @@ class IndirectVectorReader implements IVectorReader {
     @Override
     public VectorReader select(int startIdx, int len) {
         return select(IntStream.range(startIdx, startIdx + len).toArray());
+    }
+
+    @Override
+    public @NotNull VectorReader openSlice(@NotNull BufferAllocator al) {
+        return new IndirectVectorReader(reader.openSlice(al), indirection);
     }
 
     @Override

